File: CodexTempo/CodexUsageReader.cs
using System.IO;
using System.Text;
using System.Text.Json;

namespace CodexTempo;

public sealed class CodexUsageReader : IDisposable
{
    private readonly string _sessionsRoot;
    private readonly FileSystemWatcher? _watcher;
    private readonly object _gate = new();
    private readonly Dictionary<string, CacheEntry> _cache = new(StringComparer.OrdinalIgnoreCase);
    private string? _latestCandidate;
    private DateOnly? _baselineDate;
    private long _baselineReset;
    private double _baselineWeekUsed;
    private sealed record CacheEntry(long FileLength, UsageSnapshot Snapshot);

    public CodexUsageReader()
    {
        var codexHome = CodexPathResolver.ResolveHome();
        _sessionsRoot = Path.Combine(codexHome, "sessions");

        if (Directory.Exists(_sessionsRoot))
        {
            _watcher = new FileSystemWatcher(_sessionsRoot, "*.jsonl")
            {
                IncludeSubdirectories = true,
                NotifyFilter = NotifyFilters.FileName | NotifyFilters.LastWrite | NotifyFilters.CreationTime,
                EnableRaisingEvents = true
            };
            _watcher.Created += Track;
            _watcher.Changed += Track;
            _watcher.Renamed += (_, e) => SetCandidate(e.FullPath);
        }
    }

    private void Track(object sender, FileSystemEventArgs e) => SetCandidate(e.FullPath);

    private void SetCandidate(string path)
    {
        lock (_gate) _latestCandidate = path;
    }

    public async Task<UsageSnapshot?> ReadLatestAsync(CancellationToken cancellationToken = default)
    {
        if (!Directory.Exists(_sessionsRoot)) return null;

        string? candidate;
        lock (_gate) candidate = _latestCandidate;

        var files = new List<FileInfo>();
        if (candidate is not null && File.Exists(candidate))
            files.Add(new FileInfo(candidate));

        // Only enumerate the current and previous date folders. FileSystemWatcher
        // catches newly-created sessions between refreshes.
        foreach (var day in new[] { DateTime.Today, DateTime.Today.AddDays(-1) })
        {
            var dir = Path.Combine(_sessionsRoot, day.ToString("yyyy"), day.ToString("MM"), day.ToString("dd"));
            if (!Directory.Exists(dir)) continue;
            files.AddRange(new DirectoryInfo(dir).EnumerateFiles("*.jsonl"));
        }

        UsageSnapshot? newest = null;
        var now = DateTimeOffset.Now;
        LimitWindow? five = null;
        DateTimeOffset fiveAt = DateTimeOffset.MinValue;
        LimitWindow? week = null;
        DateTimeOffset weekAt = DateTimeOffset.MinValue;

        foreach (var file in files.DistinctBy(f => f.FullName).OrderByDescending(f => f.LastWriteTimeUtc).Take(8))
        {
            cancellationToken.ThrowIfCancellationRequested();
            var snapshot = await ReadFileAsync(file.FullName, cancellationToken);
            if (snapshot is null) continue;
            if (newest is null || snapshot.CapturedAt > newest.CapturedAt) newest = snapshot;
            if (snapshot.FiveHour is { } shortWindow && shortWindow.ResetsAt > now && snapshot.CapturedAt > fiveAt)
                (five, fiveAt) = (snapshot.FiveHour, snapshot.CapturedAt);
            if (snapshot.Week is { } weekWindow && weekWindow.ResetsAt > now && snapshot.CapturedAt > weekAt)
                (week, weekAt) = (snapshot.Week, snapshot.CapturedAt);
        }

        if (newest is null) return null;
        SetCandidate(newest.SourceFile);
        var todayUsed = await EstimateTodayUsedAsync(week, cancellationToken);
        return newest with { FiveHour = five, Week = week, TodayUsedPercent = todayUsed };
    }

    private async Task<double?> EstimateTodayUsedAsync(LimitWindow? currentWeek, CancellationToken ct)
    {
        if (currentWeek is null) return null;
        var today = DateOnly.FromDateTime(DateTime.Today);
        if (_baselineDate == today && _baselineReset == currentWeek.ResetsAt.ToUnixTimeSeconds())
            return Math.Max(0, currentWeek.UsedPercent - _baselineWeekUsed);

        UsageSnapshot? baseline = null;
        var previousDay = DateTime.Today.AddDays(-1);
        var previousDir = Path.Combine(_sessionsRoot, previousDay.ToString("yyyy"),
            previousDay.ToString("MM"), previousDay.ToString("dd"));
        if (Directory.Exists(previousDir))
        {
            foreach (var file in new DirectoryInfo(previousDir).EnumerateFiles("*.jsonl")
                         .OrderByDescending(f => f.LastWriteTimeUtc).Take(6))
            {
                var snapshot = await ReadFileAsync(file.FullName, ct);
                if (snapshot?.Week?.ResetsAt == currentWeek.ResetsAt &&
                    (baseline is null || snapshot.CapturedAt > baseline.CapturedAt))
                    baseline = snapshot;
            }
        }

        if (baseline is null)
        {
            var currentDir = Path.Combine(_sessionsRoot, DateTime.Today.ToString("yyyy"),
                DateTime.Today.ToString("MM"), DateTime.Today.ToString("dd"));
            if (Directory.Exists(currentDir))
            {
                foreach (var file in new DirectoryInfo(currentDir).EnumerateFiles("*.jsonl")
                             .OrderBy(f => f.CreationTimeUtc).Take(16))
                {
                    var snapshot = await ReadOldestFileSnapshotAsync(file.FullName, ct);
                    if (snapshot?.Week?.ResetsAt == currentWeek.ResetsAt &&
                        (baseline is null || snapshot.CapturedAt < baseline.CapturedAt))
                        baseline = snapshot;
                }
            }
        }

        _baselineDate = today;
        _baselineReset = currentWeek.ResetsAt.ToUnixTimeSeconds();
        _baselineWeekUsed = baseline?.Week?.UsedPercent ?? currentWeek.UsedPercent;
        return currentWeek.UsedPercent >= _baselineWeekUsed
            ? currentWeek.UsedPercent - _baselineWeekUsed
            : currentWeek.UsedPercent;
    }

    private static async Task<UsageSnapshot?> ReadOldestFileSnapshotAsync(string path, CancellationToken ct)
    {
        const int probeSize = 512 * 1024;
        try
        {
            await using var stream = new FileStream(path, FileMode.Open, FileAccess.Read,
                FileShare.ReadWrite | FileShare.Delete, 8192, FileOptions.Asynchronous);
            var head = await ReadRangeAsync(stream, 0, Math.Min(stream.Length, probeSize), ct);
            return ParseOldest(head, path);
        }
        catch (IOException) { }
        catch (UnauthorizedAccessException) { }
        return null;
    }

    private async Task<UsageSnapshot?> ReadFileAsync(string path, CancellationToken ct)
    {
        const int probeSize = 512 * 1024;
        try
        {
            await using var stream = new FileStream(path, FileMode.Open, FileAccess.Read,
                FileShare.ReadWrite | FileShare.Delete, 8192, FileOptions.Asynchronous | FileOptions.SequentialScan);

            CacheEntry? cached;
            lock (_gate) _cache.TryGetValue(path, out cached);
            if (cached is not null && cached.FileLength == stream.Length)
                return cached.Snapshot;

            if (cached is not null && cached.FileLength <= stream.Length)
            {
                // Read only data appended since the last pass, with a small overlap
                // in case the previous pass ended while a JSONL line was mid-write.
                var start = Math.Max(0, cached.FileLength - 8192);
                var appended = await ReadRangeAsync(stream, start, stream.Length, ct);
                var newer = ParseNewest(appended, path);
                var result = newer is not null && newer.CapturedAt >= cached.Snapshot.CapturedAt
                    ? newer : cached.Snapshot;
                lock (_gate) _cache[path] = new(stream.Length, result);
                return result;
            }

            // First encounter: inspect a bounded tail, then a bounded head. Rate
            // snapshots normally appear in one of these regions. Never walk an
            // entire long transcript during widget startup.
            var tailStart = Math.Max(0, stream.Length - probeSize);
            var tail = await ReadRangeAsync(stream, tailStart, stream.Length, ct);
            var snapshot = ParseNewest(tail, path);
            if (snapshot is null && tailStart > 0)
            {
                var headEnd = Math.Min(stream.Length, probeSize);
                var head = await ReadRangeAsync(stream, 0, headEnd, ct);
                snapshot = ParseNewest(head, path);
            }
            if (snapshot is not null)
            {
                lock (_gate) _cache[path] = new(stream.Length, snapshot);
                return snapshot;
            }
        }
        catch (IOException) { }
        catch (UnauthorizedAccessException) { }
        return null;
    }

    private static async Task<string> ReadRangeAsync(FileStream stream, long start, long end, CancellationToken ct)
    {
        stream.Seek(start, SeekOrigin.Begin);
        var length = checked((int)(end - start));
        var buffer = new byte[length];
        var read = 0;
        while (read < length)
        {
            var count = await stream.ReadAsync(buffer.AsMemory(read, length - read), ct);
            if (count == 0) break;
            read += count;
        }
        return Encoding.UTF8.GetString(buffer, 0, read);
    }

    private static UsageSnapshot? ParseNewest(string text, string path)
    {
        var lines = text.Split('\n', StringSplitOptions.RemoveEmptyEntries);
        for (var i = lines.Length - 1; i >= 0; i--)
        {
            var line = lines[i];
            var jsonStart = line.IndexOf('{');
            if (jsonStart < 0 || !line.AsSpan(jsonStart).Contains("\"rate_limits\"", StringComparison.Ordinal))
                continue;

            var snapshot = ParseLine(line[jsonStart..], path);
            if (snapshot is not null) return snapshot;
        }
        return null;
    }

    private static UsageSnapshot? ParseOldest(string text, string path)
    {
        foreach (var line in text.Split('\n', StringSplitOptions.RemoveEmptyEntries))
        {
            var jsonStart = line.IndexOf('{');
            if (jsonStart < 0 || !line.AsSpan(jsonStart).Contains("\"rate_limits\"", StringComparison.Ordinal))
                continue;
            var snapshot = ParseLine(line[jsonStart..], path);
            if (snapshot is not null) return snapshot;
        }
        return null;
    }

    private static UsageSnapshot? ParseLine(string json, string path)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (!root.TryGetProperty("payload", out var payload)) return null;
            JsonElement limits;
            if (payload.TryGetProperty("rate_limits", out var directLimits))
                limits = directLimits;
            else if (payload.TryGetProperty("info", out var info) &&
                     info.TryGetProperty("rate_limits", out var nestedLimits))
                limits = nestedLimits;
            else
                return null;
            if (limits.ValueKind != JsonValueKind.Object) return null;

            // Codex can emit separate metered buckets for specific models.
            // Only the canonical "codex" bucket represents the general
            // 5-hour/weekly allowance shown by this widget.
            if (limits.TryGetProperty("limit_id", out var limitIdElement))
            {
                var limitId = limitIdElement.GetString();
                if (!string.IsNullOrWhiteSpace(limitId) &&
                    !limitId.Equals("codex", StringComparison.OrdinalIgnoreCase))
                    return null;
            }

            LimitWindow? five = null;
            LimitWindow? week = null;
            foreach (var name in new[] { "primary", "secondary" })
            {
                if (!limits.TryGetProperty(name, out var item) || item.ValueKind != JsonValueKind.Object)
                    continue;
                if (!item.TryGetProperty("used_percent", out var used) ||
                    !item.TryGetProperty("window_minutes", out var window) ||
                    !item.TryGetProperty("resets_at", out var reset)) continue;

                var value = new LimitWindow(used.GetDouble(), window.GetInt32(),
                    DateTimeOffset.FromUnixTimeSeconds(reset.GetInt64()));
                if (value.WindowMinutes is >= 270 and <= 330) five = value;
                else if (value.WindowMinutes is >= 9000 and <= 11000) week = value;
            }

            if (five is null && week is null) return null;
            var captured = root.TryGetProperty("timestamp", out var stamp) &&
                           DateTimeOffset.TryParse(stamp.GetString(), out var parsed)
                ? parsed : new FileInfo(path).LastWriteTimeUtc;
            return new UsageSnapshot(five, week, captured, path);
        }
        catch (JsonException) { return null; }
    }

    public static bool RunSelfTest()
    {
        const string canonical = """
            {"timestamp":"2026-07-30T14:40:28Z","payload":{"rate_limits":{
              "limit_id":"codex","primary":{"used_percent":31,"window_minutes":10080,"resets_at":1785903281}
            }}}
            """;
        const string modelSpecific = """
            {"timestamp":"2026-07-30T14:40:38Z","payload":{"rate_limits":{
              "limit_id":"codex_bengalfox","limit_name":"GPT-5.3-Codex-Spark",
              "primary":{"used_percent":0,"window_minutes":10080,"resets_at":1786027234}
            }}}
            """;

        var accepted = ParseLine(canonical, "canonical.jsonl");
        var rejected = ParseLine(modelSpecific, "model-specific.jsonl");
        return accepted?.Week?.UsedPercent == 31 && rejected is null;
    }

    public void Dispose() => _watcher?.Dispose();
}

File: macOS/Sources/CodexTempoMac/CodexUsageReader.swift
import Foundation

final class CodexUsageReader: @unchecked Sendable {
    private let sessionsRoot: URL
    private let fileManager: FileManager
    private let calendar: Calendar
    private let now: () -> Date
    private let probeSize = 512 * 1_024
    private var baselineDay: Date?
    private var baselineReset: Date?
    private var baselineWeekUsed: Double?

    init(
        sessionsRoot: URL = CodexPathResolver.resolveHome().appendingPathComponent("sessions", isDirectory: true),
        fileManager: FileManager = .default,
        calendar: Calendar = .current,
        now: @escaping () -> Date = Date.init
    ) {
        self.sessionsRoot = sessionsRoot
        self.fileManager = fileManager
        self.calendar = calendar
        self.now = now
    }

    func readLatest() async -> UsageSnapshot? {
        let current = now()
        let files = candidateFiles(around: current)
        var newest: UsageSnapshot?
        var fiveHour: (LimitWindow, Date)?
        var week: (LimitWindow, Date)?

        for file in files.prefix(128) {
            guard !Task.isCancelled, let snapshot = readNewest(from: file) else { continue }
            if newest == nil || snapshot.capturedAt > newest!.capturedAt { newest = snapshot }
            if let window = snapshot.fiveHour, window.resetsAt > current,
               fiveHour == nil || snapshot.capturedAt > fiveHour!.1 {
                fiveHour = (window, snapshot.capturedAt)
            }
            if let window = snapshot.week, window.resetsAt > current,
               week == nil || snapshot.capturedAt > week!.1 {
                week = (window, snapshot.capturedAt)
            }
        }

        guard let latest = newest else { return nil }
        let todayUsed = week.map { estimateTodayUsed(currentWeek: $0.0, at: current) }
        return UsageSnapshot(
            fiveHour: fiveHour?.0,
            week: week?.0,
            capturedAt: latest.capturedAt,
            source: latest.source,
            todayUsedPercent: todayUsed
        )
    }

    func estimateTodayUsed(for currentWeek: LimitWindow) async -> Double? {
        estimateTodayUsed(currentWeek: currentWeek, at: now())
    }

    private func estimateTodayUsed(currentWeek: LimitWindow, at date: Date) -> Double {
        let currentDay = calendar.startOfDay(for: date)
        if let baselineDay, calendar.isDate(baselineDay, inSameDayAs: currentDay),
           let baselineReset, sameReset(baselineReset, currentWeek.resetsAt),
           let baselineWeekUsed {
            return currentWeek.usedPercent >= baselineWeekUsed
                ? currentWeek.usedPercent - baselineWeekUsed
                : currentWeek.usedPercent
        }

        var baseline: UsageSnapshot?
        if let previousDay = calendar.date(byAdding: .day, value: -1, to: date) {
            for file in files(in: directory(for: previousDay))
                .sorted(by: { modificationDate($0) > modificationDate($1) })
                .prefix(16) {
                guard let snapshot = readNewest(from: file),
                      let week = snapshot.week,
                      sameReset(week.resetsAt, currentWeek.resetsAt) else { continue }
                if baseline == nil || snapshot.capturedAt > baseline!.capturedAt {
                    baseline = snapshot
                }
            }
        }

        if baseline == nil {
            for file in files(in: directory(for: date))
                .sorted(by: { modificationDate($0) < modificationDate($1) })
                .prefix(32) {
                guard let snapshot = readOldest(from: file),
                      let week = snapshot.week,
                      sameReset(week.resetsAt, currentWeek.resetsAt) else { continue }
                if baseline == nil || snapshot.capturedAt < baseline!.capturedAt {
                    baseline = snapshot
                }
            }
        }

        let used = baseline?.week?.usedPercent ?? currentWeek.usedPercent
        baselineDay = currentDay
        baselineReset = currentWeek.resetsAt
        baselineWeekUsed = used
        return currentWeek.usedPercent >= used
            ? currentWeek.usedPercent - used
            : currentWeek.usedPercent
    }

    private func candidateFiles(around date: Date) -> [URL] {
        var result: [URL] = []
        for offset in [0, -1] {
            guard let day = calendar.date(byAdding: .day, value: offset, to: date) else { continue }
            result.append(contentsOf: files(in: directory(for: day)))
        }
        return result.sorted { modificationDate($0) > modificationDate($1) }
    }

    private func directory(for date: Date) -> URL {
        let components = calendar.dateComponents([.year, .month, .day], from: date)
        return sessionsRoot
            .appendingPathComponent(String(format: "%04d", components.year ?? 0), isDirectory: true)
            .appendingPathComponent(String(format: "%02d", components.month ?? 0), isDirectory: true)
            .appendingPathComponent(String(format: "%02d", components.day ?? 0), isDirectory: true)
    }

    private func files(in directory: URL) -> [URL] {
        (try? fileManager.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ))?.filter { $0.pathExtension == "jsonl" } ?? []
    }

    private func modificationDate(_ url: URL) -> Date {
        (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
    }

    private func readNewest(from url: URL) -> UsageSnapshot? {
        guard let data = boundedData(from: url) else { return nil }
        for line in data.tail.split(separator: 0x0A).reversed() {
            if let parsed = Self.parseLine(Data(line), source: url.path, fallbackDate: modificationDate(url)) {
                return parsed
            }
        }
        guard let head = data.head else { return nil }
        for line in head.split(separator: 0x0A).reversed() {
            if let parsed = Self.parseLine(Data(line), source: url.path, fallbackDate: modificationDate(url)) {
                return parsed
            }
        }
        return nil
    }

    private func readOldest(from url: URL) -> UsageSnapshot? {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }
        let head = (try? handle.read(upToCount: probeSize)) ?? Data()
        for line in head.split(separator: 0x0A) {
            if let parsed = Self.parseLine(Data(line), source: url.path, fallbackDate: modificationDate(url)) {
                return parsed
            }
        }
        return nil
    }

    private func boundedData(from url: URL) -> (tail: Data, head: Data?)? {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }
        guard let size = try? handle.seekToEnd() else { return nil }
        let tailStart = size > UInt64(probeSize) ? size - UInt64(probeSize) : 0
        try? handle.seek(toOffset: tailStart)
        let tail = (try? handle.readToEnd()) ?? Data()
        guard tailStart > 0 else { return (tail, nil) }
        try? handle.seek(toOffset: 0)
        return (tail, (try? handle.read(upToCount: probeSize)) ?? Data())
    }

    static func parseLine(_ data: Data, source: String, fallbackDate: Date) -> UsageSnapshot? {
        guard let text = String(data: data, encoding: .utf8),
              let start = text.firstIndex(of: "{"),
              text[start...].contains("\"rate_limits\""),
              let jsonData = String(text[start...]).data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
              let payload = root["payload"] as? [String: Any] else { return nil }

        let limits: [String: Any]?
        if let direct = payload["rate_limits"] as? [String: Any] {
            limits = direct
        } else if let info = payload["info"] as? [String: Any] {
            limits = info["rate_limits"] as? [String: Any]
        } else {
            limits = nil
        }
        guard let limits else { return nil }
        if let limitID = limits["limit_id"] as? String,
           !limitID.isEmpty,
           limitID.caseInsensitiveCompare("codex") != .orderedSame {
            return nil
        }

        let windows = parseWindows(
            limits,
            usedKey: "used_percent",
            durationKey: "window_minutes",
            resetKey: "resets_at"
        )
        guard windows.five != nil || windows.week != nil else { return nil }
        let captured = (root["timestamp"] as? String).flatMap(parseDate) ?? fallbackDate
        return UsageSnapshot(
            fiveHour: windows.five,
            week: windows.week,
            capturedAt: captured,
            source: source
        )
    }

    static func parseWindows(
        _ limits: [String: Any],
        usedKey: String,
        durationKey: String,
        resetKey: String
    ) -> (five: LimitWindow?, week: LimitWindow?) {
        var five: LimitWindow?
        var week: LimitWindow?
        for name in ["primary", "secondary"] {
            guard let item = limits[name] as? [String: Any],
                  let used = number(item[usedKey]),
                  let duration = number(item[durationKey]),
                  let reset = number(item[resetKey]) else { continue }
            let window = LimitWindow(
                usedPercent: used,
                windowMinutes: Int(duration),
                resetsAt: Date(timeIntervalSince1970: reset)
            )
            if (270...330).contains(window.windowMinutes) { five = window }
            if (9_000...11_000).contains(window.windowMinutes) { week = window }
        }
        return (five, week)
    }

    private static func number(_ value: Any?) -> Double? {
        if let number = value as? NSNumber { return number.doubleValue }
        if let string = value as? String { return Double(string) }
        return nil
    }

    private static func parseDate(_ value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value)
    }

    private func sameReset(_ left: Date, _ right: Date) -> Bool {
        abs(left.timeIntervalSince(right)) <= 90
    }
}
